### ISSUE TITLE
QPainter error

### ISSUE THREAD

**Issue by jhennawi** [ejeschke/ginga#692]
Hi All, 

I'm getting this QPainter error out put which causes channels to crash. The context is that I'm plotting images to multiple channels using the RC interface, and using ch.cut_levels() to manually change the
cuts of a given channel. 

This crash strangely goes away when I take out the cut levels call. I can try to create a script that reproduces the error, but any thoughts of the top of your head? This is pretty strange. 

QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::setRenderHint: Painter must be active to set rendering hints
QPainter::setPen: Painter not active
QPainter::setBrush: Painter not active
QPainter::pen: Painter not active
QPainter::setPen: Painter not active
QPainter::setFont: Painter not active
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::setRenderHint: Painter must be active to set rendering hints
QPainter::setPen: Painter not active
QPainter::setBrush: Painter not active
QPainter::pen: Painter not active
QPainter::setPen: Painter not active
QPainter::setFont: Painter not active
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::setRenderHint: Painter must be active to set rendering hints
QPainter::setPen: Painter not active
QPainter::setBrush: Painter not active
QPainter::pen: Painter not active
QPainter::setPen: Painter not active
QPainter::setFont: Painter not active
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::setRenderHint: Painter must be active to set rendering hints
QPainter::setPen: Painter not active
QPainter::setBrush: Painter not active
QPainter::pen: Painter not active
QPainter::setPen: Painter not active
QPainter::setFont: Painter not active
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::begin: A paint device can only be painted by one painter at a time.
QPainter::setRenderHint: Painter must be active to set rendering hints
QPainter::setPen: Painter not active
QPainter::setBrush: Painter not active
QPainter::pen: Painter not active
QPainter::setPen: Painter not active
QPainter::setFont: Painter not active

**Comment by pllim** [comment]
I encountered it back in #112 for a different use case. This is probably a different cause since @ejeschke patched that one like 3 years ago. However, I was able to debug it by inserting print statements in suspected places and running that debug version of Ginga to reproduce the errors. Maybe Eric J has a better idea.

**Comment by ejeschke** [comment]
This is probably due to the remote call actually invoking a Qt graphics operation, which is a no-no. Pretty easily solved using either `gui_do()` or `gui_call()`, but probably should be done in the general case from the `RC` plugin.  I'll take a look.

**Comment by ejeschke** [comment]
@jhennawi, are you creating channels dynamically, or do the channels exist before you are doing the operations on them?

**Comment by jhennawi** [comment]
Hi @ejeschke, 

I'm creating channels dynamics, loading images in them, and then trying to modify cut levels. The sequence of commands looks something like this, but this is executed multiple times for different images and different channels. 

viewer = grc.RemoteClient('localhost', 9000)
ch = viewer.channel(chname)
ch.load_np(chname, img, 'fits', header)
ch.cut_levels(cut_min, cut_max)

**Comment by ejeschke** [comment]
In that case, I think I may see the problem.  I'm not sure, but I suspect there may be a race condition in the calls, in that the channel is not completely set up and loaded before the cut levels is trying to do its work.

What happens if you put a `time.sleep(2)` between the load and the `cut_levels()` call? Not advocating this as a solution, necessarily, but want to test my hypothesis.

**Comment by jhennawi** [comment]
Yeah I will try this and report back. 

Also, can you look at my question on issue #693, I can understand what you mean by "ginga" handle.

Many Thanks!
Joe

**Comment by jhennawi** [comment]
Hi @ejeschke, 

The sleep command did not solve it. I even tried 5s. 

What is really weird, is that it only crapps somehow if I set the cuts in the first image I load. For subsequent images it does not have a problem with me setting the cuts. So if I just don't set the
cuts on the first image it works fine. 

Joe

**Comment by pllim** [comment]
Did the first image fall of out of data cache? (Just wildly guessing. I don't use RC.)

**Comment by ejeschke** [comment]
> Did the first image fall of out of data cache?

But `cut_levels()` only changes the cut levels for the currently displayed image.  It shouldn't matter that the first image got pushed out, as long as there *is* an image in the channel.

**Comment by ejeschke** [comment]
> ... but this is executed multiple times for different images and different channels.

What if you add a sleep between the set of calls for each channel being modified?

When I follow your example above for a single image it is no problem to cut levels.

**Comment by jhennawi** [comment]
Hi @ejeschke, 

Adding the sleep command did not fix the issue, please see above.

**Comment by jhennawi** [comment]
Hi @ejeschke, 

The QPainter issue appears to have gone away. I've since then rebuilt my ginga because of some updates to python packages, but I cannot really say whether that is what fixed it. 

But I stil find weird behavior when cuts are adjusted for certain images using the RC interface. For example, in one case the WCS matching appears to stop working for the images for which I've adjusted cuts. Whereas when I don't adjust the cuts for these two images, then the WCS matching works. 

The weird thing is that in the code I adjust cuts of two images (albeit to the same values) and it does not cause problems. If you want to look at the code it is here, but it is not so straightforward. 

https://github.com/pypeit/PypeIt/blob/master/pypeit/scripts/show_twod.py

In the meantime, I can try to create a simple example that reproduces the bug.
